### PR TITLE
[go_router] Use the correct configuration to build the state passed to the `onExit`

### PR DESCRIPTION
**Pull request by ValentinVignal**
While working on https://github.com/flutter/packages/pull/6614, I notice some issues with the state that is given to the `onExit` (see the run https://github.com/flutter/packages/pull/6614/checks?check_run_id=24284539541)

This PR uses the correct configuration to build the state and pass it to the `onExit`

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/go_router/lib/src/delegate.dart [hunk 1]
```diff
@@ -233,7 +233,7 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
           exitingMatches.length - 1,
           context: navigatorContext,
           matches: exitingMatches,
-          configuration: configuration,
+          // configuration: configuration,
```
Comment (chunhtai) [inline packages/go_router/lib/src/delegate.dart:236]: ?!

File: packages/go_router/lib/src/delegate.dart [hunk 2]
```diff
@@ -276,15 +274,14 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
...
-          configuration: configuration,
         );
       }
       return SynchronousFuture<bool>(false);
     }
 
     final FutureOr<bool> exitFuture = goRoute.onExit!(
       context,
-      match.buildState(_configuration, configuration),
+      match.buildState(_configuration, currentConfiguration),
```
Comment (chunhtai) [inline packages/go_router/lib/src/delegate.dart:284]: so the difference is whether we use the new RouteMatchList after the change vs old RouteMatchList?

I think for redirect we use the new RouteMatchList. it seems to previous implementation is more inline with the redirect API no?

Comment (chunhtai) [reply packages/go_router/lib/src/delegate.dart:284]: I maybe wrong here. do you have use case that using the old routematchList is more preferred? on the other hand, I plan to redesign redirect, so maybe we should just provide both before and after?

File: packages/go_router/lib/src/delegate.dart (hunk 1 above)
Comment (ValentinVignal) [reply packages/go_router/lib/src/delegate.dart]: Oh my ! Sorry about that. I removed it in [refactor: Remove commented code](https://github.com/flutter/packages/pull/6623/commits/560d3f44054795cc828101f3ad889830979d9853)

File: packages/go_router/lib/src/delegate.dart (hunk 2 above)
Comment (ValentinVignal) [reply packages/go_router/lib/src/delegate.dart:283]: Yes, I found one in https://github.com/flutter/packages/pull/6614. I was making `go_router_builder` use `onExit` in some tests failed https://github.com/flutter/packages/pull/6614/checks?check_run_id=24284539541.

The issue was when using `.go()`. The `onExit` had the state of the new page, not the previous page.

The test I wrote in this PR 

```dart
'It should provide the correct path parameters to the onExit callback during a go
```

fails without my changes.

> On the other hand, I plan to redesign redirect, so maybe we should just provide both before and after?

Sure, do you want me to change something then?

Comment (chunhtai) [reply packages/go_router/lib/src/delegate.dart:283]: I will need to redesign the onExit in the future. It also has issue with popscope and cupertinobackswipe. Thing may become a lot different after I resolve all these issues.

After thinking about it more, I think we should just keep it simple for now. I think what you have is fine.

**Review by chunhtai — APPROVED** [review]
LGTM

**Comment by ValentinVignal** [comment]
@hangyujin Is there something I need to do about this PR ?

**Review by hangyujin — APPROVED** [review]
LGTM

**Pull request merged**